python/ray/data/datasource/parquet_datasource.py

```diff
@@ -158,6 +158,30 @@ def set_schema_pickled(self, schema_pickled: bytes):
         self.schema_pickled = schema_pickled
 
 
+def _check_for_legacy_tensor_type(schema):
+    """Check for the legacy tensor extension type and raise an error if found.
+
+    Ray Data uses an extension type to represent tensors in Arrow tables. Previously,
+    the extension type extended `PyExtensionType`. However, this base type can expose
+    users to arbitrary code execution. To prevent this, we don't load the type by
+    default.
+    """
+    import pyarrow as pa
+
+    for name, type in zip(schema.names, schema.types):
+        if isinstance(type, pa.UnknownExtensionType) and isinstance(
+            type, pa.PyExtensionType
+        ):
+            raise RuntimeError(
+                f"Ray Data couldn't infer the type of column '{name}'. This might mean "
+                "you're trying to read data written with an older version of Ray. "
+                "Reading data written with older versions of Ray might expose you to "
+                "arbitrary code execution. To try reading the data anyway, set "
+                "`RAY_DATA_AUTOLOAD_PYEXTENSIONTYPE=1` on all nodes."
+                "To learn more, see https://github.com/ray-project/ray/issues/41314."
```

Comment:
I find this message slightly confusing (or underselling the risk). It's not only data "written with older versions of Ray" that might expose the user, it's reading any random Parquet file that will expose the user _if_ you enabled this `RAY_DATA_AUTOLOAD_PYEXTENSIONTYPE=1` and have imported Ray